src/main/kotlin/org/jitsi/utils/stats/RateTracker.kt

```diff
@@ -0,0 +1,156 @@
+/*
+* Copyright @ 2018 - present 8x8, Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+package org.jitsi.utils.stats
+
+import org.jitsi.utils.ms
+import java.lang.IllegalArgumentException
+import java.time.Clock
+import java.time.Duration
+
+open class RateTracker @JvmOverloads constructor(
+    /**
+     * The number of buckets.
+     */
+    private val numBuckets: Int,
+    /**
+     * The size of each bucket in milliseconds.
+     */
+    private val bucketSizeMs: Int = 1,
+    private val clock: Clock = Clock.systemUTC()
+) {
+    constructor(
+        /**
+         * The duration of the window for which values will be kept.
+         */
+        windowSize: Duration,
+        /**
+         * The duration of each bucket in the window. This must divide [windowSize] evenly.
+         */
+        bucketSize: Duration = 1.ms,
+        clock: Clock = Clock.systemUTC()) : this(
+        numBuckets = (windowSize.toMillis() / bucketSize.toMillis()).toInt(),
+        bucketSizeMs = bucketSize.toMillis().toInt(),
+        clock = clock) {
+
+        if (bucketSizeMs * numBuckets.toLong() != windowSize.toMillis()) {
+            throw IllegalArgumentException(
+                "The bucketSize (${bucketSize.toMillis()} ms) must divide the window size " +
+                    "(${windowSize.toMillis()} ms) evenly.")
+        }
+    }
+    /**
+     * Total count recorded in buckets.
+     */
+    private var accumulatedCount: Long = 0
+
+    /**
+     * Counters are kept in buckets (circular buffer), with one bucket per [bucketSizeMs] milliseconds.
+     */
+    private val buckets: LongArray = LongArray(numBuckets + 1)
+
+    /**
+     * Bucket index of oldest counter recorded in buckets.
+     */
+    private var oldestIndex = 0
+
+    /**
+     * Oldest time recorded in buckets.
+     */
+    private var oldestTime: Long = 0
+
+    /**
+     * The size of the window in milliseconds.
+     */
+    val windowSizeMs = numBuckets * bucketSizeMs
+
+    /**
+     * Convert a clock time to the local time representation (one tick is [bucketSizeMs] milliseconds).
+     */
+    private fun coerceMs(timeMs: Long) = timeMs / bucketSizeMs
+
+    @Synchronized
+    private fun eraseOld(now: Long) {
+        val newOldestTime = now - buckets.size + 1
```

Comment:
I'll probably have to go over all of this some more, but this is confusing to read: we subtract the number of buckets from a timestamp and get a timestamp?

Reply:
Yes, because the timestamp is in "ticks per bucketSizeMs". Note this class is a translation of `RateStatistics` with this feature (bucket size > 1) added on top.